### PR TITLE
Add support for LLVM 12

### PR DESCRIPTION
**Pull request by QuLogic**
Some cross-ports from LLVM's Go bindings directory, plus some compat code to work on older LLVM.

Passes tests on LLVM 12 and 11, but I haven't yet tested with tinygo.

### CODE REVIEW

**Comment by QuLogic** [comment]
Testing with tinygo, this seems to pass the smoke tests, though it seems like wasm might be broken.

File: backports.cpp [hunk 1]
```diff
@@ -62,3 +63,9 @@ LLVMMetadataRef LLVMGoDIBuilderCreateCompileUnit(
     SplitNameLen, Kind, DWOId, SplitDebugInlining, DebugInfoForProfiling);
 #endif
 }
+
+void LLVMGoAddConstantPropagationPass(LLVMPassManagerRef PM) {
+#if LLVM_VERSION_MAJOR < 12
+  LLVMAddConstantPropagationPass(PM);
+#endif
+}
```
Comment (aykevl) [inline backports.cpp:71]: I think this function should be removed entirely and update TinyGo to not use the constant propagation pass (I believe the SCCP pass is a good replacement).

**Comment by aykevl** [comment]
You will also need to add support for the new file paths in LLVM 12. For example, take a look at 70c558526021f5f1880fa083f5f6950831c07d4c. It needs llvm_config_linux_llvm12.go and a `!llvm12` in llvm_config_linux.go.

File: backports.cpp (hunk 1 above)
Comment (QuLogic) [reply backports.cpp]: OK, did that here and in `tinygo`.

Comment (QuLogic) [reply backports.cpp]: The LLVM12 release notes are (partially) out now, and they say to use `InstSimplify`; should we do that instead?

Comment (aykevl) [reply backports.cpp]: Oh, interesting. Might be worth trying.

However, I've replaced this pass with IPSCCP so this isn't necessary for the switch to LLVM 12.

**Comment by foutrelis** [comment]
Hi, is this in a state that can be merged?

@anatol Ideally we want to keep tinygo on system LLVM but please let me know if you can push a static build to Arch to unblock the LLVM 12 rebuilds. 🐈

**Comment by aykevl** [comment]
This PR looks good so could (in my opinion) be merged already, even if LLVM 12 support isn't there yet in TinyGo.

**Comment by aykevl** [comment]
@foutrelis note that more work is required to port TinyGo over to LLVM 12, it won't just work by bumping the LLVM version.

**Comment by QuLogic** [comment]
Sure, I don't think there's anything here that _needs_ a working tinygo before merging.

**Comment by QuLogic** [comment]
Ping? I see I can merge this, but nobody's approved it or anything...

**Comment by deadprogram** [comment]
Since @aykevl said it looks good, going to merge now. Thanks @QuLogic for working on this!

**Pull request merged**